lib/streamlit/web/server/memory_media_file_storage.py

```diff
@@ -0,0 +1,156 @@
+    mimetype
+        Any string. Will be converted to bytes and used to compute a hash.
+    filename
+        Any string. Will be converted to bytes and used to compute a hash.
+    """
+    filehash = hashlib.new("sha224")
+    filehash.update(data)
+    filehash.update(bytes(mimetype.encode()))
+
+    if filename is not None:
+        filehash.update(bytes(filename.encode()))
+
+    return filehash.hexdigest()
+
+
+def _get_extension_for_mimetype(mimetype: str) -> str:
+    # Python mimetypes preference was changed in Python versions, so we specify
+    # a preference first and let Python's mimetypes library guess the rest.
+    # See https://bugs.python.org/issue4963
+    #
+    # Note: Removing Python 3.6 support would likely eliminate this code
+    if mimetype in PREFERRED_MIMETYPE_EXTENSION_MAP:
+        return PREFERRED_MIMETYPE_EXTENSION_MAP[mimetype]
+
+    extension = mimetypes.guess_extension(mimetype)
+    if extension is None:
+        return ""
+
+    return extension
+
+
+class MemoryFile(NamedTuple):
+    content: bytes
+    mimetype: str
+    filename: Optional[str]
+
+
+class MemoryMediaFileStorage(MediaFileStorage, CacheStatsProvider):
+    def __init__(self, media_endpoint: str):
+        self._files_by_id: Dict[str, MemoryFile] = {}
+        self._media_endpoint = media_endpoint
+
+    def load_and_get_id(
+        self,
+        path_or_data: Union[str, bytes],
+        mimetype: str,
+        filename: Optional[str] = None,
+    ) -> str:
+        """Add a file to the manager and return its ID."""
+        file_data: bytes
+        if isinstance(path_or_data, str):
+            file_data = self._read_file(path_or_data)
+        else:
+            file_data = path_or_data
+
+        # Because our file_ids are stable, if we already have a file with the
+        # given ID, we don't need to create a new one.
+        file_id = _calculate_file_id(file_data, mimetype, filename)
+        media_file = self._files_by_id.get(file_id)
+        if media_file is None:
+            LOGGER.debug("Adding media file %s", file_id)
+            media_file = MemoryFile(
+                content=file_data, mimetype=mimetype, filename=filename
+            )
+            self._files_by_id[file_id] = media_file
+
+        return file_id
+
+    def get_file(self, file_id: str) -> MemoryFile:
+        """Return the MemoryFile with the given ID. Raise KeyError if not found."""
+        return self._files_by_id[file_id]
+
+    def get_url(self, file_id: str) -> str:
+        """Get a URL for a given media file. Raise a MediaFileStorageError if
+        no such file exists.
+        """
+        try:
+            media_file = self._files_by_id[file_id]
+        except KeyError as e:
+            raise MediaFileStorageError(f"No media file with id '{file_id}'") from e
+
+        extension = _get_extension_for_mimetype(media_file.mimetype)
+        return f"{self._media_endpoint}/{file_id}{extension}"
```

Comment:
It feels a bit weird to me that these two functions are very structurally similar (`get_url` basically just calls `get_file` and does some string formatting on the return value) but raise different types of errors.

Would it make sense to either have `KeyErrors` raised by both or to make `MediaFileStorageErrors` more granular so that some of them can act like `KeyErrors` (probably by inheriting from `KeyError` so that `except KeyError` clauses can catch them)?

Reply:
Yeah - I prefer having MediaFileStorageError be the base class (and so not inherit from KeyError), but I agree with re-raising a KeyError as a MediaFileStorageError